### ISSUE TITLE
The builtin version of the AF_Packet plugin is missing its `zeekctl` plugin

### ISSUE THREAD

**Issue by J-Gras** [zeek/zeek#2792]
```
Warning: ignoring unrecognized node config option 'af_packet_fanout_id' given for node 'worker-1'
Warning: ignoring unrecognized node config option 'af_packet_fanout_mode' given for node 'worker-1'
Warning: ignoring unrecognized node config option 'af_packet_buffer_size' given for node 'worker-1'
```
AF_Packet lists the plugin together with other dist files: https://github.com/zeek/zeek-af_packet-plugin/blob/master/CMakeLists.txt#L14 Not sure how much of a hack this is anyways.

**Comment by awelzel** [comment]
Hmm, I'm not sure what the right answer is. 

> Not sure how much of a hack this is anyways.

Seems to have worked as intended before, so...

Specifically for the problem at hand, we could merge `af_packet.py` into `zeekctl`. Now that AF_Packet can be considered part of Zeek that doesn't seem overly bad, even if it voids the modularity thought.

The broader topic would be if `zeek_plugin_dist_files` for builtin plugins should somehow make it into the Zeek installation tree as this currently doesn't happen.

**Comment by J-Gras** [comment]
> Specifically for the problem at hand, we could merge af_packet.py into zeekctl. Now that AF_Packet can be considered part of Zeek that doesn't seem overly bad, even if it voids the modularity thought.

This might be problematic if people disable the builtin plugin and install the external plugin.

As a workaround, maybe the file can be moved "manually" to the correct place somewhere here: https://github.com/zeek/zeek/blob/master/CMakeLists.txt#L740

**Comment by awelzel** [comment]
> This might be problematic if people disable the builtin plugin and install the external plugin.

The external plugin would not ship the af_packet.py file anymore.

**Comment by ckreibich** [comment]
> The broader topic would be if zeek_plugin_dist_files for builtin plugins should somehow make it into the Zeek installation tree as this currently doesn't happen.

Yeah, that's what breaks this. I think the proper fix is to make Python scripts their own thing in the cmake plugin logic (like BiFs, BinPAC files, etc). This would gain us:
- Separation of Python scripts from other dist stuff like VERSION, README, etc
- An opportunity to keep the Zeek install tree cleaner, since we have a dedicated place for Python in the `lib64/zeek/python` folder. This already has a `zeekctl/plugin` subdirectory, but the dist mechanism doesn't use it — zeekctl scans the whole Zeek plugin install tree for .py files.
- Generalization to arbitrary Python functionality, which also opens up extensibility of other tooling like zkg.

If we do this, it might be best to wait for zeek/cmake#60 to land, since it revamps how these sets of files get expressed.

**Comment by J-Gras** [comment]
> The external plugin would not ship the af_packet.py file anymore.

This would mean the latest plugin version is missing zeekctl support when installed to older Zeek versions. If I remember correctly, supporting older Zeek versions was one of the goals wrt. the external plugin.

> I think the proper fix is to make Python scripts their own thing in the cmake plugin logic (like BiFs, BinPAC files, etc).

Agreed. I assume this will eventually break compatibility as well.

If the Zeek version is available in CMake, maybe the plugin itself can be made backwards compatible with either approach. Or we drop backwards compatibility at this point...

**Comment by ckreibich** [comment]
A few ideas for things we could do right away, including in the next 5.0.x patch release:
- We could try and see what happens if we simply add `af_packet.py` to zeekctl. For folks that stick with the bundled plugin, it should just work. If somebody opts to `--disable-af-packet`, but then installs the package, we'd have two instances of the module, so one of the imports should blow up, trigger a warning, and move on.
- We could include `af_packet.py` in zeekctl, and check the cmake-level `DISABLE_AF_PACKET` bit in it. If set, skip the module's installation. 
- We could special-case the cmake logic in zeek-af_packet-plugin so it installs `af_packet.py` "manually".

**Comment by awelzel** [comment]
> A few ideas for things we could do right away, including in the next 5.0.x patch release:

5.0 doesn't ship af_packet. That only came with 5.2. We'll learn something from adding it to a 5.0 specific zeekctl branch, but not sure it's worth it, actually (we can't bump 5.0 to zeekctl master due to breaking changes).

For 5.2, shipping af_packet.py with zeekctl seems a reasonable fix (cc @timwoj). I'll prep a PR.

IMO we should consider Zeek 5.2 + external AF_Packet to be "on-your-own" territory. If there's a zeekctl warning, then there's a warning. I for one still hope we'll go the [proposed spicy-plugin route of merging, too](https://github.com/zeek/spicy-plugin/issues/183) ;-)

**Comment by awelzel** [comment]
PR over in zeekctl: https://github.com/zeek/zeekctl/pull/47

**Comment by timwoj** [comment]
Yep, that should work fine.

**Comment by initconf** [comment]
> IMO we should consider Zeek 5.2 + external AF_Packet to be "on-your-own" territory. If there's a zeekctl warning, then there's a warning. I for one still hope we'll go the [proposed spicy-plugin route of merging, too](https://github.com/zeek/spicy-plugin/issues/183) ;-)

Actually there is one category - which is not quite looked at/considered: the
upgrade to 5.2.x when previous versions were running external plugin.

Lets say there is already external AF_Packet running on 5.1.0 or 4.x.x etc - now I upgrade to 5.2.x which is missing af_packet.py (or likes of it). 

So here comes the problem - current run is already setup with node.cfg to
include 

af_packet_fanout_id=23
af_packet_fanout_mode=AF_Packet::FANOUT_HASH
af_packet_buffer_size=128*1024*1024

but 5.2.x built-in plugin doesn't support this.

Now I as a user don't even realize there is already a new af_packet plugin built
into zeek-5.2.x  but what I see is - oh this used to work and now isn't ?

I am happy to just move to built-in plugin (I'd rather do that) but there is
surprise bit enabled here and that confuses.

It further confuses that do above three settings even have meaning or ok to
comment that out ? 

So in short on-your-own is when there is a clear indication of issue one is
facing and then they get to make a choice of external-plugin or built-in.
"on-your-own" while unclear what has changed without being aware is hard place
to leave someone.

Aashish

> --

**Comment by awelzel** [comment]
Thanks for your thoughts @initconf !

> Actually there is one category - which is not quite looked at/considered: the
upgrade to 5.2.x when previous versions were running external plugin.

TLDR - if zeekctl is upgraded as well, this should now continue to function. It's essentially what Jan pointed out.

> Lets say there is already external AF_Packet running on 5.1.0 or 4.x.x etc - now I upgrade to 5.2.x which is missing af_packet.py (or likes of it).
>
> So here comes the problem - current run is already setup with node.cfg to
include
> ...
> but 5.2.x built-in plugin doesn't support this.
>
> Now I as a user don't even realize there is already a new af_packet plugin built
into zeek-5.2.x  but what I see is - oh this used to work and now isn't ?

If as a user you upgraded to Zeek 5.2 with a builtin plugin enabled *and* upgraded to the zeekctl version that comes with Zeek 5.2, the options will now be recognized as af_packet.py functionality is now included in zeekctl directly. This should solve the issue that Jan found pointed out in this ticket. The upside of violating the modularity here a bit, you can now say `lb_method = af_packet` in node.cfg.

There's still the following gap:
* User using older version of zeekctl that does not support af_packet.py, but Zeek 5.2 with builtin af-packet. But we can point this out in NEWS and ask for upgrading zeekctl.

I'll try to come up with a paragraph for Zeek 5.2 news to point out the older zeekctl + builtin af_packet version.

**Comment by initconf** [comment]
Aha! Thanks for the clear explanation Anre.

 Now it does makes sense.  Also I am so used to bundled code that idea of
not upgrading zeekctl along with rest didn’t cross my mind.  In that case
sure I am with “on-your-own” setup, esp since that appears to be a much
more custom deployment.

**Comment by timwoj** [comment]
Was this fixed by https://github.com/zeek/zeek/pull/2826?

**Comment by awelzel** [comment]
> Was this fixed by https://github.com/zeek/zeek/pull/2826?

Yes.
